File: leetcode/hard/44_isMatch.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function(s, p) {
    const cache = new Map();

    const _isMatch = (y, x) => {
        if (cache.has([s.length, p.length].join('_')))
            return cache.get([s.length, p.length].join('_'));

        const key = [y, x].join('_');

        if (cache.has(key))
            return cache.get(key);

        if (x == p.length && y == s.length) {
            cache.set(key, true);
            return true;
        }

        if (x == p.length && y < s.length) {
            cache.set(key, false);
            return false;
        }

        if (x < p.length && y == s.length) {
            for (i = x; i < p.length; ++i) {
                if (p[i] != '*') {
                    cache.set([y, i].join('_'), false);
                    cache.set(key, false);
                    return false;
                }
            }
            cache.set([p.length, s.length].join('_'), true);
            cache.set(key, true);
            return true;
        }

        if (s[y] == p[x] || p[x] == '?') {
            const r = _isMatch(y + 1, x + 1);
            cache.set(key, r)
            return r;
        }

        if (p[x] == '*') {
            let i = y
            while (i <= s.length) {
                const r = _isMatch(i, x + 1);
                if (r == true) {
                    cache.set(key, true);
                    return true;
                }
                i += 1
            }
            cache.set(key, false);
            return false;
        }

        cache.set(key, false);
        return false
    };

    return _isMatch(0, 0);
};


const test = () => {
    const params = [
        {
            'input': [
                "aababbbbaaabaabaabbbbbaabbaabaaaabaaabbbaaaaabbbaaaabaaababbbbbbabbbabaababbbaaaaaaabaaaabbbaabbbaaabaaaababbababbaabaaaaabaaababbaababbabbbbabaabababbabbabbababbbbaaaabbbaabbaabbaabababbbbaaaabbabaaabbab",
                "*bbb*b*****a*abaab****a****b***a*ab*bb***b**bb*b*aab*aaa*a*b*bbbb*a*a*****ba**bb*b*****b*a*bb*******aa"
            ],
            'output': false,
        },
        {
            'input': [
                "bbbbaababbbbaabaaabbbbbbabbabbaaabaabbbaaaaaababababbbbbaabbaababbbbbabbababbaabaaabbbababaabbabbbaabbbabbbaabbabbbabbbbabbaabbbbbbaabababbaaaababbbaaabbbbaaabbbbabaabaaababbabbbabaaabbbbbbbbaabaabbabb",
                "b*ab*b****b*a**b**b****abbba**a*baa****b*ab****bbabaaaab***ab****aba***a******aa*ba*bba****aa******b*b**",
            ],
            'output': false,
        },

        {
            'input': ["adceb", "*a*b"],
            'output': true,
        },
        {
            'input': ["aa", "a"],
            'output': false,
        },
        {
            'input': ["aa", "*"],
            'output': true,
        },
        {
            'input': ["cb", "?a"],
            'output': false,
        },
        {
            'input': ["aa", "a?"],
            'output': true,
        },
        {
            'input': ["cb", "??"],
            'output': true,
        },
        {
            'input': ["caaaaaab", "c*b"],
            'output': true,
        },
        {
            'input': ["cb", "c*b"],
            'output': true,
        },
        {
            'input': ["cbbbbbbb", "c*b"],
            'output': true,
        },
    ];

    params.forEach(({input, output}) => {
        const result = isMatch(...input);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', JSON.stringify(output),
            'result', JSON.stringify(result),
        );
    });
};

test();

File: leetcode/hard/44_isMatch.py
from typing import List
import json
from collections import deque

class Solution:
    def isMatch_0(self, s: str, p: str) -> bool:
        lengthP = len(p)
        lengthS = len(s)
        cache = {}

        res = False
        def isMatch(y, x):
            if (lengthS, lengthP) in cache:
                return cache[(lengthS, lengthP)]
            if (y, x) in cache:
                return cache[(y, x)]

            if x == lengthP and y == lengthS:
                cache[(y, x)] = True
                return True

            if x == lengthP and y < lengthS:
                cache[(y, x)] = False
                return False

            if x < lengthP and y == lengthS:
                for i in range(x, lengthP):
                    if p[i] != '*':
                        cache[(y, x)] = False
                        return False
                cache[(y, x)] = True
                return True

            if s[y] == p[x] or p[x] == '?':
                r = isMatch(y + 1, x + 1)
                cache[(y, x)] = r
                return r

            if p[x] == '*':
                i = y
                while i <= lengthS:
                    r = isMatch(i, x + 1)
                    if cache[(i, x+1)] == True:
                        cache[(y, x)] = r
                        return True
                    i += 1

                cache[(y, x)] = False
                return False

            cache[(y, x)] = False
            return False

        return isMatch(0, 0)

    def isMatch(self, s: str, p: str) -> bool:
        lengthP = len(p)
        lengthS = len(s)
        cache = {}

        res = False
        def isMatch(y, x):
            if (lengthS, lengthP) in cache:
                return cache[(lengthS, lengthP)]
            if (y, x) in cache:
                return cache[(y, x)]

            if x == lengthP and y == lengthS:
                cache[(y, x)] = True
                return True

            if x == lengthP and y < lengthS:
                cache[(y, x)] = False
                return False

            if x < lengthP and y == lengthS:
                for i in range(x, lengthP):
                    if p[i] != '*':
                        cache[(y, x)] = False
                        return False
                cache[(y, x)] = True
                return True

            if s[y] == p[x] or p[x] == '?':
                r = isMatch(y + 1, x + 1)
                cache[(y, x)] = r
                return r

            if p[x] == '*':
                cache[(y, x)] = isMatch(y, x + 1) or isMatch(y + 1, x)
                return cache[(y, x)]

            cache[(y, x)] = False
            return False

        return isMatch(0, 0)


def test ():
    params = [
        # {
        #     'input': [
        #         "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        #         "*aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa*"
        #     ],
        #     'output': False,
        # },
        {
            'input': [
                "aababbbbaaabaabaabbbbbaabbaabaaaabaaabbbaaaaabbbaaaabaaababbbbbbabbbabaababbbaaaaaaabaaaabbbaabbbaaabaaaababbababbaabaaaaabaaababbaababbabbbbabaabababbabbabbababbbbaaaabbbaabbaabbaabababbbbaaaabbabaaabbab",
                "*bbb*b*****a*abaab****a****b***a*ab*bb***b**bb*b*aab*aaa*a*b*bbbb*a*a*****ba**bb*b*****b*a*bb*******aa"
            ],
            'output': False,
        },
        {
            'input': [
                "bbbbaababbbbaabaaabbbbbbabbabbaaabaabbbaaaaaababababbbbbaabbaababbbbbabbababbaabaaabbbababaabbabbbaabbbabbbaabbabbbabbbbabbaabbbbbbaabababbaaaababbbaaabbbbaaabbbbabaabaaababbabbbabaaabbbbbbbbaabaabbabb",
                "b*ab*b****b*a**b**b****abbba**a*baa****b*ab****bbabaaaab***ab****aba***a******aa*ba*bba****aa******b*b**",
            ],
            'output': False,
        },

        {
            'input': ["adceb", "*a*b"],
            'output': True,
        },
        {
            'input': ["aa", "a"],
            'output': False,
        },
        {
            'input': ["aa", "*"],
            'output': True,
        },
        {
            'input': ["cb", "?a"],
            'output': False,
        },
        {
            'input': ["aa", "a?"],
            'output': True,
        },
        {
            'input': ["cb", "??"],
            'output': True,
        },
        {
            'input': ["caaaaaab", "c*b"],
            'output': True,
        },
        {
            'input': ["cb", "c*b"],
            'output': True,
        },
        {
            'input': ["cbbbbbbb", "c*b"],
            'output': True,
        },
    ]
    solution = Solution()

    for param in params:
        s, p = param['input']
        result = solution.isMatch(s, p)
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()
